File: static/js/capacity.js
// Capacity Planning page JS
let capacityData = null;

function pageInit() {
    loadCapacity();
}

async function loadCapacity() {
    const roleFilter = document.getElementById('capacity-role-filter')?.value || '';
    let url = `${API_URL}/capacity`;
    if (roleFilter) url += `?role=${roleFilter}`;
    try {
        capacityData = await fetch(url).then(r => r.json());
    } catch (e) { console.error('Capacity load failed:', e); return; }

    // Populate role filter dropdown (only on first load)
    const filterEl = document.getElementById('capacity-role-filter');
    if (filterEl && filterEl.options.length <= 1 && capacityData.roles) {
        const roleLabels = { admin: 'Admin', sm_specialist: 'SMM', designer: 'Designer', motion_designer: 'Motion', moderator: 'Moderator', copywriter: 'Writer' };
        capacityData.roles.forEach(r => {
            const opt = document.createElement('option');
            opt.value = r;
            opt.textContent = roleLabels[r] || r;
            filterEl.appendChild(opt);
        });
    }

    renderRoleSummary(capacityData.role_summary);
    renderCapacityBars(capacityData.capacity_bars);
    renderHeatmap(capacityData.heatmap);
    renderUnassigned(capacityData.unassigned);
    renderDeadlines(capacityData.deadlines);
}

function filterCapacityByRole() {
    loadCapacity();
}

const ROLE_LABELS = { admin: 'Admin', sm_specialist: 'SMM Specialist', designer: 'Designer', motion_designer: 'Motion Designer', moderator: 'Moderator', copywriter: 'Copywriter' };
const ROLE_COLORS = { admin: 'border-red-400 bg-red-50', sm_specialist: 'border-green-400 bg-green-50', designer: 'border-purple-400 bg-purple-50', motion_designer: 'border-orange-400 bg-orange-50', moderator: 'border-blue-400 bg-blue-50', copywriter: 'border-yellow-400 bg-yellow-50' };
const ROLE_ICONS = { admin: 'fa-shield-halved text-red-500', sm_specialist: 'fa-bullhorn text-green-500', designer: 'fa-paintbrush text-purple-500', motion_designer: 'fa-film text-orange-500', moderator: 'fa-user-check text-blue-500', copywriter: 'fa-pen-nib text-yellow-500' };

function renderRoleSummary(summary) {
    const el = document.getElementById('role-summary');
    if (!el || !summary?.length) { if (el) el.innerHTML = ''; return; }
    el.innerHTML = summary.map(rs => {
        const color = ROLE_COLORS[rs.role] || 'border-gray-400 bg-gray-50';
        const icon = ROLE_ICONS[rs.role] || 'fa-user text-gray-500';
        const utilColor = rs.utilization > 80 ? 'text-red-600' : rs.utilization > 50 ? 'text-amber-600' : 'text-green-600';
        return `<div class="rounded-xl p-4 border-l-4 ${color}">
            <div class="flex items-center gap-2 mb-2"><i class="fa-solid ${icon}"></i><span class="text-xs font-bold">${ROLE_LABELS[rs.role] || rs.role}</span></div>
            <div class="text-2xl font-bold ${utilColor}">${rs.utilization}%</div>
            <div class="text-[10px] text-gray-500">${rs.users} members · ${rs.active}/${rs.required} tasks</div>
        </div>`;
    }).join('');
}

function renderCapacityBars(bars) {
    const el = document.getElementById('capacity-bars');
    if (!el) return;
    if (!bars?.length) { el.innerHTML = '<p class="text-gray-400 text-center py-4">No team members found</p>'; return; }

    el.innerHTML = bars.map(b => {
        const pct = Math.min(b.utilization, 100);
        const barColor = pct > 80 ? 'bg-red-500' : pct > 50 ? 'bg-amber-500' : 'bg-green-500';
        const roleLabel = ROLE_LABELS[b.role] || b.role;
        const icon = ROLE_ICONS[b.role] || 'fa-user text-gray-500';
        return `<div class="flex items-center gap-3">
            <div class="w-32 flex-shrink-0">
                <div class="flex items-center gap-2"><i class="fa-solid ${icon} text-xs"></i><span class="text-sm font-semibold truncate">${esc(b.username)}</span></div>
                <span class="text-[10px] text-gray-400">${roleLabel}</span>
            </div>
            <div class="flex-1">
                <div class="w-full bg-gray-100 rounded-full h-4 relative overflow-hidden">
                    <div class="${barColor} h-4 rounded-full transition-all duration-500" style="width:${pct}%"></div>
                    <span class="absolute inset-0 flex items-center justify-center text-[10px] font-bold ${pct > 50 ? 'text-white' : 'text-gray-600'}">${b.active}/${b.required}</span>
                </div>
            </div>
            <div class="w-14 text-right text-sm font-bold ${pct > 80 ? 'text-red-600' : pct > 50 ? 'text-amber-600' : 'text-green-600'}">${b.utilization}%</div>
        </div>`;
    }).join('');
}

function renderHeatmap(heatmap) {
    const el = document.getElementById('workload-heatmap');
    if (!el || !heatmap?.length) { if (el) el.innerHTML = '<p class="text-gray-400 text-center py-4">No data</p>'; return; }

    const dates = heatmap[0]?.days?.map(d => d.date) || [];
    const today = new Date().toISOString().split('T')[0];

    let html = '<table class="w-full text-xs" style="min-width:600px"><thead><tr><th class="text-left py-1 pr-3 text-gray-500 font-semibold" style="width:120px">Team Member</th>';
    dates.forEach(d => {
        const short = d.substring(5);
        const dayName = new Date(d + 'T12:00:00').toLocaleDateString('en-US', { weekday: 'short' });
        const isToday = d === today;
        html += `<th class="text-center py-1 px-1 ${isToday ? 'text-indigo-600 font-bold' : 'text-gray-400'}" style="min-width:36px"><div>${dayName}</div><div>${short}</div></th>`;
    });
    html += '</tr></thead><tbody>';

    heatmap.forEach(row => {
        html += `<tr><td class="py-1 pr-3 font-medium truncate">${esc(row.username)}</td>`;
        row.days.forEach(d => {
            const c = d.count;
            let bg = 'bg-gray-50';
            if (c >= 5) bg = 'bg-red-500 text-white';
            else if (c >= 3) bg = 'bg-amber-400 text-white';
            else if (c >= 1) bg = 'bg-green-400 text-white';
            html += `<td class="text-center py-1 px-1"><div class="w-8 h-8 mx-auto rounded flex items-center justify-center text-xs font-semibold ${bg}" title="${d.date}: ${c} posts">${c || ''}</div></td>`;
        });
        html += '</tr>';
    });
    html += '</tbody></table>';

    // Legend
    html += '<div class="flex items-center gap-3 mt-3 text-[10px] text-gray-500"><span>Load:</span><span class="flex items-center gap-1"><span class="w-4 h-4 rounded bg-gray-50 border"></span>0</span><span class="flex items-center gap-1"><span class="w-4 h-4 rounded bg-green-400"></span>1-2</span><span class="flex items-center gap-1"><span class="w-4 h-4 rounded bg-amber-400"></span>3-4</span><span class="flex items-center gap-1"><span class="w-4 h-4 rounded bg-red-500"></span>5+</span></div>';

    el.innerHTML = html;
}

function renderUnassigned(items) {
    const el = document.getElementById('unassigned-list');
    if (!el) return;
    if (!items?.length) { el.innerHTML = '<p class="text-gray-400 text-center py-4"><i class="fa-solid fa-circle-check text-green-400 text-2xl mb-2 block"></i>All work is assigned</p>'; return; }

    el.innerHTML = items.map(item => {
        const statusLabels = { pending_review: 'Needs Reviewer', in_design: 'Needs Designer', approved: 'Needs SM', scheduled: 'Needs SM' };
        const statusLabel = statusLabels[item.workflow_status] || item.workflow_status;
        const priBadge = item.priority === 'urgent' ? 'bg-red-100 text-red-700' : item.priority === 'high' ? 'bg-orange-100 text-orange-700' : 'bg-gray-100 text-gray-600';
        const topic = getTopicPreview(item.topic, 40) || 'Untitled';
        return `<a href="/calendar?open_post=${item.id}" class="flex items-center gap-3 p-3 bg-gray-50 rounded-lg hover:bg-gray-100 transition">
            <span class="px-2 py-0.5 rounded text-[10px] font-bold ${priBadge}">${item.priority || 'normal'}</span>
            <div class="min-w-0 flex-1">
                <p class="text-sm font-semibold truncate">${esc(topic)}</p>
                <p class="text-[10px] text-gray-400">${esc(item.client_name || '')} · ${getPlatformIcon(item.platforms)} ${esc(item.post_type || '')}</p>
            </div>
            <span class="px-2 py-0.5 rounded text-[10px] font-semibold bg-red-50 text-red-600 whitespace-nowrap">${statusLabel}</span>
        </a>`;
    }).join('');
}

function renderDeadlines(items) {
    const el = document.getElementById('deadlines-list');
    if (!el) return;
    if (!items?.length) { el.innerHTML = '<p class="text-gray-400 text-center py-4">No upcoming deadlines</p>'; return; }

    el.innerHTML = items.map(item => {
        const dt = new Date(item.scheduled_at);
        const dateStr = dt.toLocaleDateString('en-US', { weekday: 'short', month: 'short', day: 'numeric' });
        const timeStr = dt.toLocaleTimeString('en-US', { hour: 'numeric', minute: '2-digit' });
        const assignee = item.sm_name || item.designer_name || 'Unassigned';
        const topic = getTopicPreview(item.topic, 40) || 'Untitled';
        const isUrgent = (dt - new Date()) < 86400000; // less than 24h
        return `<a href="/calendar?open_post=${item.id}" class="flex items-center gap-3 p-3 ${isUrgent ? 'bg-red-50 border border-red-200' : 'bg-gray-50'} rounded-lg hover:bg-gray-100 transition">
            <div class="text-center flex-shrink-0 ${isUrgent ? 'text-red-600' : 'text-indigo-600'}">
                <div class="text-lg font-bold">${dt.getDate()}</div>
                <div class="text-[10px]">${dt.toLocaleDateString('en-US', { month: 'short' })}</div>
            </div>
            <div class="min-w-0 flex-1">
                <p class="text-sm font-semibold truncate">${esc(topic)}</p>
                <p class="text-[10px] text-gray-400">${esc(item.client_name || '')} · ${timeStr} · ${getPlatformIcon(item.platforms)}</p>
            </div>
            <span class="text-[10px] text-gray-500 flex-shrink-0"><i class="fa-solid fa-user text-indigo-400"></i> ${esc(assignee)}</span>
        </a>`;
    }).join('');
}

File: routes/capacity.py
import json
from datetime import datetime, timedelta
from flask import Blueprint, jsonify, request
from models import get_db, dict_from_row, dicts_from_rows

capacity_bp = Blueprint('capacity', __name__)


@capacity_bp.route('/api/capacity', methods=['GET'])
def get_capacity():
    """Return team capacity data: heatmap, bars, role summary, unassigned, deadlines."""
    db = get_db()
    role_filter = request.args.get('role', '')
    now = datetime.now()
    month_start = now.strftime('%Y-%m-01')

    # Get all active users
    user_query = "SELECT id, username, role, job_title FROM users WHERE is_active=1"
    user_params = []
    if role_filter:
        user_query += " AND role=?"
        user_params = [role_filter]
    users = dicts_from_rows(db.execute(user_query, user_params).fetchall())

    # Get all clients with content requirements
    clients = dicts_from_rows(db.execute(
        "SELECT id, name, content_requirements, assigned_designer_id, assigned_motion_id, assigned_sm_id, assigned_writer_id, assigned_manager_id FROM clients"
    ).fetchall())

    # === CAPACITY BARS ===
    capacity_bars = []
    for user in users:
        uid = user['id']
        role = user['role']

        # Count active posts assigned to this user this month (not posted/failed)
        active_count = 0
        if role in ('designer', 'motion_designer'):
            row = db.execute("""
                SELECT COUNT(*) as c FROM scheduled_posts
                WHERE (assigned_designer_id=? OR assigned_motion_id=?)
                  AND workflow_status NOT IN ('posted', 'failed')
                  AND created_at >= ?
            """, (uid, uid, month_start)).fetchone()
            active_count = row['c']
        elif role == 'sm_specialist':
            row = db.execute("""
                SELECT COUNT(*) as c FROM scheduled_posts
                WHERE (assigned_sm_id=? OR created_by_id=?)
                  AND workflow_status NOT IN ('posted', 'failed')
                  AND created_at >= ?
            """, (uid, uid, month_start)).fetchone()
            active_count = row['c']
        elif role == 'moderator':
            row = db.execute("""
                SELECT COUNT(*) as c FROM scheduled_posts
                WHERE workflow_status IN ('approved', 'scheduled')
                  AND created_at >= ?
            """, (month_start,)).fetchone()
            active_count = row['c']
        elif role == 'admin':
            row = db.execute("""
                SELECT COUNT(*) as c FROM scheduled_posts
                WHERE workflow_status NOT IN ('posted', 'failed')
                  AND created_at >= ?
            """, (month_start,)).fetchone()
            active_count = row['c']

        # Calculate required from client content_requirements
        required = 0
        for c in clients:
            assigned = False
            if role == 'designer' and c.get('assigned_designer_id') == uid:
                assigned = True
            elif role == 'motion_designer' and c.get('assigned_motion_id') == uid:
                assigned = True
            elif role == 'sm_specialist' and (c.get('assigned_sm_id') == uid or c.get('assigned_writer_id') == uid):
                assigned = True
            elif role == 'moderator':
                assigned = True

            if assigned:
                reqs = _parse_reqs(c.get('content_requirements'))
                for r in reqs:
                    if role == 'designer' and r.get('type') in ('post', 'story', 'carousel', 'banner', 'brochure'):
                        required += r.get('count', 0)
                    elif role == 'motion_designer' and r.get('type') in ('video', 'reel'):
                        required += r.get('count', 0)
                    elif role in ('sm_specialist', 'moderator', 'admin'):
                        required += r.get('count', 0)

        utilization = round((active_count / required * 100) if required > 0 else 0, 1)
        capacity_bars.append({
            'user_id': uid,
            'username': user['username'],
            'role': role,
            'job_title': user.get('job_title', ''),
            'active': active_count,
            'required': required,
            'utilization': utilization
        })

    # === HEATMAP: users x dates (next 14 days from month start) ===
    heatmap_start = now - timedelta(days=6)
    heatmap_end = now + timedelta(days=7)
    heatmap_data = []
    for user in users:
        uid = user['id']
        days = []
        for i in range(14):
            day = heatmap_start + timedelta(days=i)
            day_str = day.strftime('%Y-%m-%d')
            row = db.execute("""
                SELECT COUNT(*) as c FROM scheduled_posts
                WHERE (assigned_designer_id=? OR assigned_sm_id=? OR assigned_motion_id=? OR created_by_id=?)
                  AND DATE(COALESCE(scheduled_at, created_at)) = ?
                  AND workflow_status NOT IN ('posted', 'failed')
            """, (uid, uid, uid, uid, day_str)).fetchone()
            days.append({'date': day_str, 'count': row['c']})
        heatmap_data.append({
            'user_id': uid,
            'username': user['username'],
            'role': user['role'],
            'days': days
        })

    # === ROLE SUMMARY ===
    role_summary = {}
    for bar in capacity_bars:
        r = bar['role']
        if r not in role_summary:
            role_summary[r] = {'role': r, 'users': 0, 'active': 0, 'required': 0}
        role_summary[r]['users'] += 1
        role_summary[r]['active'] += bar['active']
        role_summary[r]['required'] += bar['required']
    for rs in role_summary.values():
        rs['utilization'] = round((rs['active'] / rs['required'] * 100) if rs['required'] > 0 else 0, 1)

    # === UNASSIGNED POSTS ===
    unassigned = dicts_from_rows(db.execute("""
        SELECT sp.id, sp.topic, sp.caption, sp.post_type, sp.platforms, sp.workflow_status,
               sp.scheduled_at, sp.priority, c.name as client_name
        FROM scheduled_posts sp
        LEFT JOIN clients c ON sp.client_id = c.id
        WHERE sp.workflow_status NOT IN ('posted', 'failed', 'draft')
          AND (
            (sp.workflow_status = 'in_design' AND sp.assigned_designer_id IS NULL)
            OR (sp.workflow_status = 'pending_review' AND sp.assigned_manager_id IS NULL)
            OR (sp.workflow_status IN ('approved', 'scheduled') AND sp.assigned_sm_id IS NULL)
          )
        ORDER BY CASE sp.priority WHEN 'urgent' THEN 0 WHEN 'high' THEN 1 ELSE 2 END, sp.created_at DESC
        LIMIT 20
    """).fetchall())

    # === UPCOMING DEADLINES (7 days) ===
    deadline_end = (now + timedelta(days=7)).strftime('%Y-%m-%d %H:%M:%S')
    deadlines = dicts_from_rows(db.execute("""
        SELECT sp.id, sp.topic, sp.post_type, sp.platforms, sp.scheduled_at, sp.workflow_status,
               sp.priority, c.name as client_name,
               u_d.username as designer_name, u_s.username as sm_name
        FROM scheduled_posts sp
        LEFT JOIN clients c ON sp.client_id = c.id
        LEFT JOIN users u_d ON sp.assigned_designer_id = u_d.id
        LEFT JOIN users u_s ON sp.assigned_sm_id = u_s.id
        WHERE sp.scheduled_at IS NOT NULL AND sp.scheduled_at != ''
          AND sp.scheduled_at <= ?
          AND sp.scheduled_at >= datetime('now')
          AND sp.workflow_status NOT IN ('posted', 'failed')
        ORDER BY sp.scheduled_at ASC
    """, (deadline_end,)).fetchall())

    # Get distinct roles for filter
    roles = list(set(u['role'] for u in users))

    db.close()
    return jsonify({
        'capacity_bars': capacity_bars,
        'heatmap': heatmap_data,
        'role_summary': list(role_summary.values()),
        'unassigned': unassigned,
        'deadlines': deadlines,
        'roles': sorted(roles)
    })


def _parse_reqs(json_str):
    if not json_str:
        return []
    try:
        return json.loads(json_str)
    except (json.JSONDecodeError, TypeError):
        return []
